sdk/core/azure-core/src/main/java/com/azure/core/util/polling/PollerFlux.java

```diff
@@ -0,0 +1,247 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.core.util.polling;
+
+import com.azure.core.util.logging.ClientLogger;
+import reactor.core.CoreSubscriber;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+
+import java.time.Duration;
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicIntegerFieldUpdater;
+import java.util.function.BiFunction;
+import java.util.function.Supplier;
+
+/**
+ * A Flux that asynchronously invokes an activation operation and poll the progress of a long-running
+ * operation until it complete.
+ *
+ * Each subscription to the PollerFlux will run it's own polling but the activation is guaranteed to be
+ * called only once across all subscription.
+ *
+ * e.g.: cancelling a long-running operation after a specified time.
+ *
+ *    PollerFlux<BlobCopyStatus, BlobInfo> cpBlobPollerFlux = asyncClient.beginBlobCopy();
+ *
+ *    cpBlobPollerFlux
+ *       .take(Duration.ofMinutes(30))
+ *       .last()
+ *       .flatMap(asyncPollResponse -> {
+ *           if (!asyncPollResponse.getStatus().isComplete()) {
+ *               return asyncPollResponse
+ *                   .cancelOperation()
+ *                   .then(Mono.error(new RuntimeException("LRO taking long time > 30 mins, operation is cancelled!")));
+ *           } else {
+ *               return Mono.just(asyncPollResponse);
+ *           }
+ *       }).block();
+ *
+ * e.g.: retrieving the final result of long-running operation after it's completion.
+ *
+ *   PollerFlux<BlobCopyStatus, BlobInfo> cpBlobPollerFlux = asyncClient.beginBlobCopy();
+ *
+ *   cpBlobPollerFlux
+ *       .last()
+ *       .flatMap(asyncPollResponse -> {
+ *           // LRO completed, retrieving final result.
+ *           return asyncPollResponse.getFinalResult();
+ *       }).block();
+ *
+ * @param <T> The type of poll response value
+ * @param <U> The type of the final result of long-running operation
+ */
+public final class PollerFlux<T, U> extends Flux<AsyncPollResponse<T, U>> {
```

Comment:
I'm not sure if this needs to be public API?

Reply:
Update: Synced with @JonathanGiles offline about possible extensibility point on `PollerFlux` e.g. support for "rehydrating Polling from an ID at a later stage" (from a totally different process/machine)